blackbird_python/blackbird/program.py

```diff
@@ -206,9 +207,26 @@ def __call__(self, **kwargs):
         if not self.parameters:
             raise ValueError("Program is not a template!")
 
-        prog = copy.copy(self)
+        prog = copy.deepcopy(self)
         prog._parameters = [] # pylint: disable=protected-access
 
+        new_kwargs = copy.deepcopy(kwargs)
+        for k, v in kwargs.items():
+            if isinstance(v, Iterable):
+                if np.ndim(v) != 2:
+                    raise ValueError("Invalid dim for free parameter provided. Must have dim 2.")
+
+                added_kwargs = {
+                    k + "_{}_{}".format(i, j): val
+                    for i, row in enumerate(v)
+                    for j, val in enumerate(row)
+                }
```

Comment:
Is this essentially turning something like `x = [[1, 2], [5, 6]]` into `x_0_0=1, x_0_1=2,...`?

Reply:
Yes, exactly! That way everything can be parsed into just a list of operation-dicts, where the parameters would be part of `op[i]["args"]` or `op[i]["kwargs"]` (as was the idea from the start). Apart from the way that Fabian is extracting the variables now, we could probably just use the operations, and metadata, as the only external connections to the Blackbird information.